DESIGN.md

```diff
@@ -0,0 +1,370 @@
+
+In GraphQL, fragments offer an explicit way to share selections.
+
+## Fragments
+
+```graphql
+query TwoHeroes {
+  r2: hero {
+    ...HeroDetails
+  }
+  luke: hero(episode: EMPIRE) {
+    ...HeroDetails
+  }
+}
+
+fragment HeroDetails on Character {
+  name
+}
+```
+
+This would generate an interface `HeroDetails` that could be reused between selections and between operations. This is also important because it allows you to write code that isn't tied to a specific query, like a UI component that knows how to display any hero's details.
+
+The simplest mapping may seem to be to let `R2` and `Luke` extend `HeroDetails`:
+
+```java
+class TwoHeroesQuery implements GraphQLQuery<TwoHeroesQuery.Data> {
+    interface Data {
+        R2 r2();
+        interface R2 extends HeroDetails {}
+        Luke luke();
+        interface Luke extends HeroDetails {}
+    }
+}
+
+interface HeroDetails {
+    String name();
+}
+```
+
+But this model gets a lot more complicated when nested selection sets are taken into account:
+
+```graphql
+query TwoHeroes {
+  r2: hero {
+    ...HeroDetails
+    friends {
+      appearsIn
+    }
+  }
+  luke: hero(episode: EMPIRE) {
+    ...HeroDetails
+    friends {
+      id
+    }
+  }
+}
+
+fragment HeroDetails on Character {
+  name
+  friends {
+    name
+  }
+}
+```
+
+This would have to generate something like:
+
+```java
+interface R2 extends HeroDetails {
+  String name();
+  List<Friend> friends();
+  interface Friend extends HeroDetails.Friend {
+    List<Episode> appearsIn();
+  }
+}
+
+interface Luke extends HeroDetails {
+  String name();
+  List<Friend> friends();
+  interface Friend extends HeroDetails.Friend {
+    String id();
+  }
+}
+
+interface HeroDetails {
+  String name();
+  List<Friend> friends();
+  interface Friend {
+    String name();
+  }
+}
+```
+
+In Swift, this ran into limitations of the type system because types of properties in protocols (the equivalent of interfaces) are invariant. So the `List<Friend> friends()` declarations in subinterfaces would be incompatible. Even if this isn't necessarily a problem in Java, a mapping like this quickly gets complicated, and there may be simpler alternatives.
+
+## Composition over inheritance
+
+The solution we picked in Apollo iOS is to use composition over inheritance, where fragment models are mapped onto separate objects:
+
+```java
+interface R2 {  
+  List<Friend> friends();
+  interface Friend {
+    List<Episode> appearsIn();
+  }
+  Fragments fragments();
+  interface Fragments {
+    HeroDetails heroDetails();
+  }
+}
+
+interface Luke {
+  List<Friend> friends();
+  interface Friend {
+    String id();
+  }
+  Fragments fragments();
+  interface Fragments {
+    HeroDetails heroDetails();
+  }
+}
+
+interface HeroDetails {
+  String name();
+  List<Friend> friends();
+  interface Friend {
+    String name();
+  }
+}
+```
```

Comment:
So to access the Luke's friend name we have to call this `luke.fragments(). heroDetails().friends()[0].name()` but at the same time to access friend id: `luke.friends()[0].id()`. So the `Character` object info for friends is spread across 2 places. From user perspective I guess it will be a little confusing, as I user I would like to work only with `luke.friends()`. Otherwise I need to put some logic to math users from 2 sources and merge them into one.